src/common/tusb_fifo.c

```diff
@@ -87,8 +87,15 @@ bool tu_fifo_config(tu_fifo_t *f, void* buffer, uint16_t depth, uint16_t item_si
 
 static inline uint16_t _ff_mod(uint16_t idx, uint16_t depth)
 {
-  while ( idx >= depth) idx -= depth;
```

Comment:
Thank you for detail explanation, hmm, somehow I remembered the while loop only need to do it maximum twice. Maybe many things has changed (my memory is fading). @PanRe since we use fifo as main buffering for audio/iso transfer, we don't have the option to go power of 2 only right ?

Reply:
That is very interesting as the while loop is to be executed twice at maximum! The fifo was designed not to need more that that and the while loop was only used to keep it general (in the first PR i did not use the while loop, rather two subtractions but @hathach prefered the while loop).

Could you check the values of idx when the slow-down happens? And depth also?

Reply:
Restricting to power of two was never an option for you as i remember ^^. It would simplify a lot of things, however, restricting to a power of two length is a hard limit and not very practical for audio applications... i would not do it unless there is a serious bug we can not solve otherwise!

Reply:
yeah, I remembered that. It is actually better to have arbitrary depth, that will allow us to fit the stack on memory constraint memory. Thanks for reminding, I need more SRAM for my brain !!! :sweat_smile: :sweat_smile:

Reply:
When you are doing so, please give all the variables of the fifo e.g. non_used_index_space etc.!

Reply:
Hmm,he may use the old code fron raspberrypi fork. There is a bug fix regarding fifo used in pico

Reply:
I think i know... maybe. There is a bug in fifo.c which i fixed within #593. In [line 564](https://github.com/hathach/tinyusb/blob/ab4d30fd6bca02c73eb9b4ff82db0b2b0f403344/src/common/tusb_fifo.c#L564) the command ` else if (count > f->depth)` is wrong! It should be `else if (n >= f->depth)`. This would explain why he needs exactly double the time at some iterations! This is a bug which I verified when i worked on #593! You may try this! :)

Reply:
It seems that your intent was to bound the idx (read and write) to 2*(_depth)-1. It isn't the case, at least on the pico branch. the read and write values are only ever incremented and they eventually wrap around their max uint16_t values. 
So the real correct fix would be to clamp their values to their mod depth.

Reply:
Ah thats it... ok i don't know the pico branch but the FIFO was designed as you mentioned it above!

Reply:
@PanRe : I don't think your PR fixed this. The read pointer and the write pointers are not limited. Your PR will make the write pointer be equal to the read pointer but it does not limit the read pointer. The root cause if that the function _tu_fifo_correct_read_pointer is not called if cnt is not higher than the depth. For my case it is never true. Somewhere,  there needs to be a check that the read and write pointers are not greater than their max values, regardless how many bytes are written or read.

Reply:
Do you use the master branch? Because there in `advance_pointer()` or [here](https://github.com/hathach/tinyusb/blob/1d20c84798305043c2783725a496070c39d2d966/src/common/tusb_fifo.c#L145) the pointers are limited to the mentioned maximum!

The fix i mentioned before fixes a different bug ^^

Reply:
This code is identical on the pico branch 
It seems that in the pico branch at least, max_pointer_idx and non_used_index_space get ovewritten to 0 after being initialized to the right values. I need to track this down further. 
I will let you know what I find.

Reply:
@phdussud please try again with the master branch, as mentioned above, there is relevant fixed for this exact queue issue for pico.

Reply:
Ok. The problem isn't with the vendor fifos but with _usbd_qdef which isn't correctly initialized. The fifo that is embedded in the _usbd_qdef has its max_pointer_idx and non_used_index_space left to 0. 
//--------------------------------------------------------------------+
// DCD Event
//--------------------------------------------------------------------+

// Event queue
// OPT_MODE_DEVICE is used by OS NONE for mutex (disable usb isr)
OSAL_QUEUE_DEF(OPT_MODE_DEVICE, _usbd_qdef, CFG_TUD_TASK_QUEUE_SZ, dcd_event_t);
static osal_queue_t _usbd_q;

Reply:
yeah, that one exactly.

Reply:
Ah, cool.